src/main/java/org/spongepowered/common/data/AbstractArchetype.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.data;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
+import net.minecraft.nbt.CompoundNBT;
+import org.spongepowered.api.CatalogType;
+import org.spongepowered.api.data.DataProvider;
+import org.spongepowered.api.data.Key;
+import org.spongepowered.api.data.persistence.DataView;
+import org.spongepowered.api.data.persistence.InvalidDataException;
+import org.spongepowered.api.data.value.Value;
+import org.spongepowered.api.world.Archetype;
+import org.spongepowered.api.world.LocatableSnapshot;
+import org.spongepowered.common.bridge.data.DataCompoundHolder;
+import org.spongepowered.common.data.holder.SpongeMutableDataHolder;
+import org.spongepowered.common.data.nbt.validation.DelegateDataValidator;
+import org.spongepowered.common.data.nbt.validation.RawDataValidator;
+import org.spongepowered.common.data.nbt.validation.ValidationType;
+import org.spongepowered.common.data.persistence.NbtTranslator;
+import org.spongepowered.common.data.provider.DataProviderLookup;
+
+import java.util.Collection;
+import java.util.Objects;
+
+public abstract class AbstractArchetype<C extends CatalogType, S extends LocatableSnapshot<S>, E> implements Archetype<S, E>,
+        SpongeMutableDataHolder {
+
+    protected final C type;
+    protected CompoundNBT data;
+    private final DataProviderLookup lookup = SpongeDataManager.getProviderRegistry().getProviderLookup(this.getClass());
```

Comment:
This can honestly be initialized statically or lazily, no? We don't need to make this call more than once for the entire VM's lifecycle.

Reply:
well at least once for every Archetype extending this